sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+
+
+## Design to implement required diagnostics
+
+### Changes to DPC++ headers
+
+With the exception of the `[[sycl::reqd_work_group_size()]]` and
+`[[sycl::reqd_sub_group_size()]]` attributes, all kernel optional features
+are associated with some device aspect.  For example, the `sycl::half` type
+is an optional feature which is only supported on devices that have the
+`aspect::fp16` aspect.  We can therefore use device aspects as a way to
+describe the set of optional features that a kernel uses (with the exception
+of the required work-group or sub-group size).
+
+As we will see later, it will be very helpful to decorate all APIs in DPC++
+headers that correspond to optional kernel features with the
+`[[sycl::requires()]]` attribute.  For example, the declaration of the
+`sycl::half` type would look like this:
+
+```
+using half [[sycl::requires(has(aspect::fp16))]] = cl::sycl::detail::half_impl::half;
+```
+
+In cases where the optional feature corresponds to use of a class (e.g.
+`sycl::atomic_ref`), the declaration can look like this:
+
+```
+template</*...*/>
+class [[sycl::requires(has(aspect::atomic64))]] atomic_ref {
+  /* ... */
+};
+```
+
+(In reality, we can use partial specialization tricks to decorate `atomic_ref`
+with the attribute only when the underlying type is 64-bits.  See ["Appendix:
+Adding an attribute to 8-byte `atomic_ref`"][4].)
+
+[4]: <#appendix-adding-an-attribute-to-8-byte-atomic_ref>
+
+In cases where the optional feature corresponds to a function, we can decorate
+the function's declaration with the attribute like so (demonstrating a
+hypothetical AMX multiplication extension):
+
+```
+[[sycl::requires(has(aspect::ext_intel_amx))]]
+void amx_multiply();
+```
+
+These attributes provide an explicit mapping between each optional kernel
+feature and its associated aspect.
+
+Unfortunately, the fundamental type `double` is also an optional kernel
+feature.  Since there is no type alias for `double`, there is no convenient
+place to add an attribute.  Instead, the front-end device compiler must behave
+as though there was an implicit `[[sycl::requires(has(aspect::fp64))]]`
+attribute for any device code that uses the `double` type.
+
+Note that the usage of `[[sycl::requires()]]` is slightly expanded here beyond
+the specified usage in the SYCL 2020 specification because we allow the
+attribute also on type alias declarations and class declarations.  If a device
+function does any of the following with a type alias or class that was so
+decorated, the function is assumed to "use the aspects" listed in the
+attribute:
+
+* Declares a variable of that type.
+* Has a formal parameter declared with that type.
+* Returns that type.
+
+This also includes any qualified version of the type.
+
+**TODO**: This language is not very precise.  The intent is to include most
+uses of the type, except for cases like `sizeof(T)` or `decltype(T)`.  Help
+appreciated on tightening the wording here.
```

Comment:
This is a load-bearing TODO because I don't know how to implement this from what's given. Why are those cases unimportant while the above cases are important (what about the type's usage actually matters)? The kinds of examples I'm wondering about are things like:
```
void func(const std::vector<MarkedType> &whatever); // Is this a bad use of the type?

template <typename Ty>
void func() {} // Ty isn't in the return type or parameter list but is in the signature, is a marked type bad?

template <typename Ty>
void func() noexcept(noexcept(Ty{})); // Is this a bad use of the type?

template <typename Ty>
struct S {
  void foo() {} // Does this "use" the marked type (it shows up in the signature)?
};

template <typename T>
struct S {
  explicit(std::is_same_v<T, Whatever>) S(const S&) = default; // How about this use?
};
```

Reply:
I'm going to use this comment thread to address all the questions about what it means to "use" a type that is marked with `[[sycl::requires]]` (or `[[__sycl::requires_type()]]`).

What I really want, of course, is to consider a type "used" only if the compiler ends up generating SPIR-V code that operates on an object of that type.  However, I need to map this back to some wording that makes sense to the C++ front end.  Here is an attempt at more precise wording, but comments are appreciated:

We consider that a device function "uses" a marked type `T` if it does any of the following:

* Defines an object (including a temporary object) of type `U`.
* Calls a "new expression" of type `U`.
* Throws an expression of type `U`.

where type `U` is any of the following:

* The type `T`.
* A cv-qualified version of type `T`.
* An array of, pointer to, or reference to type `T`.
* A type that derives from type `T`.
* A class type that contains a member object of type `T`.
* Any type that applies these rules recursively to type `T` (e.g. array of pointers to type `T`, etc.)

I'm open to using some different criteria too, if you think this is difficult for the FE to check.